File: List_1_C/List_1.scala
object List_1 {
  def flatten[T](list : List[List[T]]):List[T] = {
    if (list != null && list.nonEmpty) {
      list.head ++ flatten(list.tail)
    } else Nil
  }

  def count[T](x : T, list : List[T]):Int = {
    if (list != null && list.nonEmpty) {
      val ooz = if (x == list.head) 1 else 0
      ooz + count(x, list.tail)
    } else 0
  }

  def replicate[T](x : T, number : Int):List[T] = {
    if (number > 0) {
      List(x) ++ replicate(x, number - 1)
    } else Nil
  }

  def sqrList(list : List[Int]):List[Int] = {
    if (list != null && list.nonEmpty) {
      List(list.head*list.head) ++ sqrList(list.tail)
    } else Nil
  }

  def palindrome[T](list : List[T]):Boolean = {
    if (list == null) true
    else list == list.reverse
  }

  def listLength[T](list :List[T]):Int = {
    if (list != null && list.nonEmpty) {
      1 + listLength(list.tail)
    } else 0
  }
}

File: Semestr_3/Paradigms/List_1_C/List_1.ml
let rec flatten(llt) = (
    if llt = [] then [] 
    else List.hd llt @ flatten(List.tl llt)
);;

let rec count(x, lt) = (
    let oozf = (fun (y, z) -> if y = z then 1 else 0) in
    if lt = [] then 0
    else oozf(x, List.hd lt) + count(x, List.tl lt)
);;

let rec replicate(x, number) = (
    if number <= 0 then []
    else [x] @ replicate(x, number-1)
);;

let rec sqrList(li) = (
    if li = [] then []
    else [List.hd li * List.hd li] @ sqrList(List.tl li)
);;

let rec palindrome(lt) = (
    lt = List.rev lt
);;

let rec listLength(lt) = (
    if lt = [] then 0
    else 1 + listLength(List.tl lt)
);;
